### DISCUSSION TITLE
vrc-get-guiのメンテナを募集してます

### DISCUSSION THREAD

**Discussion by anatawa12** [vrc-get/vrc-get#1184] [General]
正直next.js等に詳しくないのである程度知識ある方にある程度メンテナンスを任せたいと思ってます

**Comment by Spokeek** [comment]
I used to be a **fullstack JS Dev**, i'll give the source a look and see if i can do things about it.
My Rust is, well, rusted, so i'll need to catch up on that too for backend dev, even if most of the work is done by the cli tool.

**Comment by Spokeek** [comment]
I'm rising this again as well, i offered to help on this.

But we never defined what we meant by maintenance. Or if we do, this is way too vague as a task.
Should we close this mainly dedicate a nextjs tag for all the issues we may have with this ?

**Comment by anatawa12** [comment]
I found that most of vrc-get-gui features / bugfixes requires both nextjs and rust so I think it's good to have some specific label for frontend only features and bugfixes. Thank you for suggestion!

**Comment by Spokeek** [comment]
Should we move this issue to a discussion then ? i don't think it makes sense to keep it, especially in the project board.

**Comment by ProffDea** [comment]
It does make sense making this issue into a discussion since maintaining the frontend can be done in any opinionated fashion.

Since no one really knows, the most I can do is point out the most glaring issues, and give solutions to the issues. Depending on the solutions, it can improve the future maintenance of the code and let contributors have a better understanding on how to contribute code to the project.

### Linter/Formatter

The frontend is lacking a formatter. I noticed the `.editorconfig` file added to the frontend last week which is a start. The formatter I would recommend for this project in particular would be [prettier](https://github.com/prettier/prettier). They actually support `.editorconfig` and will parse it. However, `.editorconfig` can be replaced by `.prettierrc.json` so it's more inline with the formatter's configs but this decision can be made up to the maintainers. I personally would prefer `.prettierrc.json` instead. Since `.eslintrc.json` already exists in the project, [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) should be included in the file for any rules that may be added into eslint in the future so the two don't conflict with each other.

For the sake of referencing options, I'll mention [biome](https://github.com/biomejs/biome) that can fit into this project since it supports JSX/TSX. However, biome is both a linter and a formatter which means it wouldn't make sense to keep the `.eslintrc.json` file and `eslint` packages in `package.json`. Biome would be used instead of prettier for formatting. I can't imagine this being a popular choice for this project because nextjs uses eslint thus a `.eslintrc.json` file gets created upon project generation, but eslint isn't properly configured at the moment. I have my bias with biome as it has sane defaults, minimal setup, and it's new compared to the alternatives so it would be awesome to see it being used in this project, but realistically I can see eslint+prettier being the preferred choice, but I could be wrong! For all I know, you could end up prefering biome over eslint+prettier. This would be my pick.

I wasn't even able to setup eslint for this project so I have no comment on eslint.

For linters like eslint and biome, I can imagine it being a tough transition after adding any rules since the project has a lot of code already and these linters are going to complain with dozens of errors. Particularly variables/functions with `any` types. This is part of the code I can see everyone agreeing on maintaining into appropriate types as we're using typescript afterall.

Maybe these linters and formatters will make your life easier to the point where a maintainer for the frontend wouldn't be required anymore. Once any of these are implemented, I won't be able to write any better code than anyone else contributing to the project.

### PNPM instead of NPM

[pnpm](https://github.com/pnpm/pnpm) has many advantages over npm that there's really no reason not to use it other than it being another dependency step to the project setup! I would highly recommend the switch to pnpm! The main reason I'll give why is because I tried building the project with pnpm before only for it to error. I've found out that the cause of the error was because `tailwind-merge` was being used as a dependency in the project without it being listed as a dependency! You can see the following git blame that I added `tailwind-merge` as a dependency. Meanwhile, the commits before mine, you can find `tailwind-merge` being used in the project.

![image](https://github.com/vrc-get/vrc-get/assets/54484875/f984f2b6-26fb-4a04-a8a2-51c618782d3e)

This code was before my commit.
https://github.com/vrc-get/vrc-get/blob/2dc7aad51ddaf6c535e472a02861c64b3e5dc6f2/vrc-get-gui/components/select.tsx#L7

This `package.json` is from the same commit as the code from above. It lacks the `tailwind-merge` dependency!
https://github.com/vrc-get/vrc-get/blob/2dc7aad51ddaf6c535e472a02861c64b3e5dc6f2/vrc-get-gui/package.json

The only reason why I caught this was because of pnpm. I don't know how this happened in the first place, but there's no reason why it should happen again, and pnpm will prevent it from happening again by telling you!

Here's what the error looks like that caught my attention on the `tailwind-merge` package.
![image](https://github.com/vrc-get/vrc-get/assets/54484875/6de68396-a1cc-463a-bdbc-3e88fe2443e4)

You can replicate this by installing pnpm and following the commands:

```bash
git clone --recurse-submodules https://github.com/vrc-get/vrc-get.git
cd vrc-get
git checkout 2dc7aad51ddaf6c535e472a02861c64b3e5dc6f2
```

Change the following in `package.json` to pnpm:
```diff
  "scripts": {
-    "tauri": "npm install && tauri",
+    "tauri": "pnpm install && tauri",
    "dev": "next dev",
-    "build": "npm install && npm run build:licenses && npm run build:next",
+    "build": "pnpm install && pnpm run build:licenses && pnpm run build:next",
    "build:next": "next build",
    "build:licenses": "node ./scripts/build-license-json.mjs",
    "start": "next start",
    "lint": "next lint"
  },
```

```bash
cd vrc-get-gui
pnpm run build
```

You should get the same error as I did once it reaches the next build step!

### Rust

This part may seem off-topic but I find this detail about the project perplexing to me. All typescript files use tabs instead of whitespaces for indentation meanwhile all rust files use whitespaces instead of tabs for indentation. I prefer tabs and it seems like this project prefers tabs too, so why not have the rest of the project use tabs for indentation to keep it consistent?

Make a `rustfmt.toml` file at the root of the project with the following contents and run `cargo fmt --all` in the project directory.

```toml
# rustfmt.toml

hard_tabs = true
```

The project has a lot of toml files, so why not format those with [taplo](https://github.com/tamasfe/taplo) too! Make a `taplo.toml` file at the root of the project with the following contents and run `taplo format -c ./taplo.toml` after installing the taplo-cli.

```toml
# taplo.toml

[formatting]
indent_string = "	"
```

Thank you for taking the time to read my comment if you made it this far down here! I seem to tend to write big comments so please ask any questions or add more context for this issue! I can't say I'd be a maintainer with the appropriate views for this project but that's because I wouldn't be confident in understanding how the project should be structured and tailored to your needs. But for the time being, I'll see where else I can apply myself to contribute more to the project! I hope any of these suggestions gets implemented for the better of the project.

**Reply by anatawa12** [reply]
Thank you for pointing out issues suggestions.

**Code formatting**

I know prettier is the popular formatting tool, and I've tried before, but I feel it's too opinionated.
I don't like prettier because it removes line breaks in method chains (or other medium length lines).
(prettier/prettier#7884 and other related issues)
I have almost no opinion on code formatting and using language or framework default.
However, there are few exceptions, and "I should have some control about line breaking based on logic meaning" is one of them.
(Another is "formatting should not force huge diff with few code changes" but prettier has very strong trust of line length limit would also make huge diff with small changes.)

The reason why I'm using tab in tsx is because Next.js generated ones use them.
In my opinion, there is no problem with using different indent for different language but if there is, I'll make use of spaces for tsx.
(AFAIK, using tab in rust is not as common as spaces in tsx)
What's the problem with mixing then in the same project?

I didn't try biome before, so It might be good to try it.

For TOML, I didn't know there were formatters for it.
However, after a little googling, it looks it doesn't support an Intellij platform (format on save), so I don't want to use it.

**Using PNPM**

The reason why I'm using npm is to reduce the cost to launch the dev version of gui for rust developers.
npm is already installed when they install Node.js, and they don't need to install another package manager.
(This is also the reason why I'm using any (latest) LTS for Node.js in this project.)
This is another reason why we have `npm install` command in `tauri` script in `package.json`.

When corepack package manager manager is stable and enabled for pnpm by default, (in other words, no extra installation is needed for pnpm) I'll use pnpm.

Of course, it's problems that have missing dependencies in package.json, but I think it's not a big deal because there is lock file, so I want to keep NPM for reasons above.

**Reply by ProffDea** [reply]
> (https://github.com/prettier/prettier/issues/7884 and other related issues)

I can see the purpose of biome existing now lol, even biome has an entire following page about the weird things with prettier
https://biomejs.dev/formatter/differences-with-prettier/

From what I've noticed, biome formats method chains onto their own new line so there's that!

If you wanna look more into biome, maybe it'd be a good idea to disable the linter and only have the formatter enabled for the time being? It would probably be ideal to have a PR that enables the linter along with all the changes to the frontend to make the linter happy. It's going to have a LOT of errors with the default recommended rules enabled lmao

> (AFAIK, using tab in rust is not as common as spaces in tsx)
What's the problem with mixing then in the same project?

None! I think it should be fine since rust will tell the IDE to use the right styling for files anyways. The official [rust styling guide](https://doc.rust-lang.org/beta/style-guide/index.html#indentation-and-line-width) seem to prefer whitespaces over tabs meanwhile the nodejs ecosystem seem to prefer tabs over whitespaces since Next.js uses tabs and biome has [tabs set by default](https://biomejs.dev/reference/configuration/#formatterindentstyle) for it's formatter. Because rust and the nodejs ecosystem have different preferrences, I can see the topic being debatable whether to pick tabs or whitespaces. The easiest choice would just to do nothing and let each have it's own styling. If a formatter gets integrated into the project for the frontend, it'll help to make sure other people's IDEs are using the right styling indentation for the files, otherwise would just have to rely on something like the `.editorconfig` to avoid easy mistakes.

> For TOML, I didn't know there were formatters for it.

I'm glad to at least spread awareness about cool stuff!

> When corepack package manager manager is stable and enabled for pnpm by default, (in other words, no extra installation is needed for pnpm) I'll use pnpm.

This makes sense. I wasn't aware about corepack, it looks interesting! I think it's fine if pnpm doesn't get integrated any time soon, it'd definitely be nice for help with maintenance later in the future.

**Comment by anatawa12** [comment]
I was considering just close & create new discussion or converting to discussion and I'm going to convert to discussion.
